src/connectedk8s/azext_connectedk8s/_clientproxyutils.py

```diff
@@ -0,0 +1,135 @@
+            if int(tup[3][1]) == int(port):
+                return True
+    except Exception as e:
+        telemetry.set_exception(exception=e, fault_type=consts.Port_Check_Fault_Type,
+                                summary='Failed to check if port is in use.')
+        if platform.system() != 'Darwin':
+            logger.info("Failed to check if port is in use. " + str(e))
+        return False
+    return False
+
+
+def close_subprocess_and_raise_cli_error(proc_subprocess, msg):
+    proc_subprocess.terminate()
+    raise CLIInternalError(msg)
+
+
+def check_if_csp_is_running(clientproxy_process):
+    if clientproxy_process.poll() is None:
+        return True
+    else:
+        return False
+
+
+def make_api_call_with_retries(uri, data, method, tls_verify, fault_type, summary, cli_error, clientproxy_process):
+    for i in range(consts.API_CALL_RETRIES):
+        try:
+            response = requests.request(method, uri, json=data, verify=tls_verify)
+            return response
+        except Exception as e:
+            time.sleep(5)
+            if i != consts.API_CALL_RETRIES - 1:
+                pass
+            else:
+                telemetry.set_exception(exception=e, fault_type=fault_type, summary=summary)
+                close_subprocess_and_raise_cli_error(clientproxy_process, cli_error + str(e))
+
+
+def fetch_pop_publickey_kid(api_server_port, clientproxy_process):
+    requestbody = {}
+    poppublickey_uri = f'https://localhost:{api_server_port}/identity/poppublickey'
+    # Needed to prevent skip tls warning from printing to the console
+    original_stderr = sys.stderr
+    f = open(os.devnull, 'w')
+    sys.stderr = f
+
+    get_publickey_response = make_api_call_with_retries(poppublickey_uri, requestbody, "get", False, consts.Get_PublicKey_Info_Fault_Type,
+                                                        'Failed to fetch public key info from clientproxy',
+                                                        "Failed to fetch public key info from client proxy", clientproxy_process)
+
+    sys.stderr = original_stderr
+    publickey_info = json.loads(get_publickey_response.text)
+    kid = publickey_info['publicKey']['kid']
+
+    return kid
+
+
+def fetch_and_post_at_to_csp(cmd, api_server_port, tenantId, kid, clientproxy_process):
+    req_cnfJSON = {"kid": kid, "xms_ksl": "sw"}
+    req_cnf = base64.urlsafe_b64encode(json.dumps(req_cnfJSON).encode('utf-8')).decode('utf-8')
+
+    # remove padding '=' character
+    if req_cnf[len(req_cnf) - 1] == '=':
+        req_cnf = req_cnf[:-1]
+
+    token_data = {"token_type": "pop", "key_id": kid, "req_cnf": req_cnf}
+    profile = Profile(cli_ctx=cmd.cli_ctx)
+    try:
+        credential, _, _ = profile.get_login_credentials(subscription_id=profile.get_subscription()["id"], resource=consts.KAP_1P_Server_App_Scope)
+        accessToken = credential.get_token(consts.KAP_1P_Server_App_Scope, data=token_data)
+        jwtToken = accessToken.token
```

Comment:
Is this supported by MSAL? Could you please add some PR description explaining how this works?

Reply:
Done, Added the detailed description

Reply:
> We have made changes in our Go application(which generates the PoP token) to accept this Jwt token and further generate PoP token using that token.

Is this implementation open-source? If so, could you link to it?

Reply:
It's not open-source. But I believe you'll have access - https://msazure.visualstudio.com/One/_git/AzureArc-GuestProxy?path=%2Fpkg%2Fadal-with-pop&version=GBmaster&_a=contents